ivy/data_classes/factorized_tensor/parafac2_tensor.py

```diff
@@ -0,0 +1,571 @@
+# local
+from .base import FactorizedTensor
+import ivy
+
+
+class Parafac2Tensor(FactorizedTensor):
+    def __init__(self, parafac2_tensor):
+        super().__init__()
+
+        shape, rank = ivy.Parafac2Tensor.validate_parafac2_tensor(parafac2_tensor)
+        weights, factors, projections = parafac2_tensor
+
+        if weights is None:
+            weights = ivy.ones(rank, dtype=factors[0].dtype)
+
+        self.shape = shape
+        self.rank = rank
+        self.factors = factors
+        self.weights = weights
+        self.projections = projections
+
+    @classmethod
+    def from_CPTensor(self, cp_tensor, parafac2_tensor_ok=False):
+        """
+        Create a Parafac2Tensor from a CPTensor.
+
+        Parameters
+        ----------
+        cp_tensor
+            CPTensor or Parafac2Tensor
+            If it is a Parafac2Tensor, then the argument
+            ``parafac2_tensor_ok`` must be True'
+        parafac2_tensor
+            Whether or not Parafac2Tensors can be used as input.
+
+        Returns
+        -------
+            Parafac2Tensor with factor matrices and weigths extracted from a CPTensor
+        """
+        if parafac2_tensor_ok and len(cp_tensor) == 3:
+            return Parafac2Tensor(cp_tensor)
+        elif len(cp_tensor) == 3:
+            raise TypeError(
+                "Input is not a CPTensor. If it is a Parafac2Tensor, then the argument"
+                " ``parafac2_tensor_ok`` must be True"
+            )
+
+        weights, (A, B, C) = cp_tensor
+        Q, R = ivy.qr(B)
+        projections = [Q for _ in range(ivy.shape(A)[0])]
+        B = R
+        return Parafac2Tensor((weights, (A, B, C), projections))
+
+    def __getitem__(self, index):
+        if index == 0:
+            return self.weights
+        elif index == 1:
+            return self.factors
+        elif index == 2:
+            return self.projections
+        else:
+            raise IndexError(
+                f"You tried to access index {index} of a PARAFAC2 tensor.\n"
+                "You can only access index 0, 1 and 2 of a PARAFAC2 tensor"
+                "(corresponding respectively to the weights, factors and projections)"
+            )
+
+    def __iter__(self):
+        yield self.weights
+        yield self.factors
+        yield self.projections
+
+    def __len__(self):
+        return 3
+
+    def __repr__(self):
+        message = (
+            f"(weights, factors, projections) : rank-{self.rank} Parafac2Tensor of"
+            f" shape {self.shape} "
+        )
```

Comment:
Let's move this after the init function and  add the `# Built-ins # ` comment just to be consistent with other factorized data classes. Could you also add a `__setitem__` method to this.